File: test/integration/endpoint_url_https_test.go
package integration_test

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"encoding/pem"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/localstack/lstk/test/integration/env"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// minimalStateZip builds a minimal valid state export ZIP, mirroring
// mockStateServer's inline construction in snapshot_save_test.go.
func minimalStateZip(t *testing.T) []byte {
	t.Helper()
	var buf bytes.Buffer
	zw := zip.NewWriter(&buf)
	f, err := zw.Create("state.json")
	require.NoError(t, err)
	_, err = f.Write([]byte(`{"services":{}}`))
	require.NoError(t, err)
	require.NoError(t, zw.Close())
	return buf.Bytes()
}

// httpsCertFileEnv writes srv's certificate to a temp PEM file and returns the
// "SSL_CERT_FILE=..." entry that makes the lstk subprocess trust it. lstk
// itself has no --insecure/trust-this-cert flag (a deliberate Non-Goal — see
// design.md), so the only way an exec'd subprocess can trust a self-signed
// httptest.NewTLSServer cert is via the OS-level trusted-root mechanism its
// TLS stack actually reads at verification time. Go's crypto/x509 honors
// SSL_CERT_FILE only on the unix builds it lists in root_unix.go (linux,
// freebsd, netbsd, openbsd, dragonfly) — darwin's cgo-based verifier goes
// through Security.framework instead and never consults it, and Windows uses
// its own store. So these tests only run on Linux; elsewhere they'd either
// skip silently or falsely fail on an OS-trust-store limitation that has
// nothing to do with the code under test.
func httpsCertFileEnv(t *testing.T, srv *httptest.Server) string {
	t.Helper()
	path := filepath.Join(t.TempDir(), "test-ca.pem")
	pemBytes := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: srv.Certificate().Raw})
	require.NoError(t, os.WriteFile(path, pemBytes, 0o600))
	return "SSL_CERT_FILE=" + path
}

func requireLinuxForSSLCertFileTrust(t *testing.T) {
	t.Helper()
	if runtime.GOOS != "linux" {
		t.Skip("SSL_CERT_FILE trust for a self-signed httptest cert is only honored by Go's Linux x509 verifier; see httpsCertFileEnv")
	}
}

// awsHealthTLSServer is awsHealthServer's https counterpart, simulating a
// LocalStack cloud-hosted ephemeral instance. It serves the same handler, so
// every route the http mock answers is answered here too.
func awsHealthTLSServer(t *testing.T) *httptest.Server {
	t.Helper()
	return httptest.NewTLSServer(awsHealthHandler())
}

// TestAWSCommandEndpointURLHTTPSNoDockerRequired mirrors
// TestAWSCommandEndpointURLNoDockerRequired but proves the resolved https
// scheme reaches the wrapped aws binary unchanged rather than being
// downgraded to http (Decision 7 / task 7.2).
func TestAWSCommandEndpointURLHTTPSNoDockerRequired(t *testing.T) {
	t.Parallel()
	requireLinuxForSSLCertFileTrust(t)

	srv := awsHealthTLSServer(t)
	defer srv.Close()

	fakeDir := writeFakeAWSEcho(t)
	e := env.With(env.DisableEvents, "1").With("PATH", fakeDir).With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost, httpsCertFileEnv(t, srv))

	stdout, stderr, err := runLstk(t, testContext(t), t.TempDir(), e, "--endpoint-url", srv.URL, "aws", "s3", "ls")
	require.NoError(t, err, "stderr: %s", stderr)
	assert.Contains(t, stdout, "ENDPOINT:"+srv.URL)
	assert.Contains(t, stdout, "ARGS:s3 ls")
}

// TestCDKCommandEndpointURLHTTPSNoDockerRequired mirrors
// TestCDKAWSEndpointURLBypassesDockerCheck's shape but for an explicit
// --endpoint-url https:// target, proving the IaC proxy layer forwards the
// real scheme to the wrapped tool's environment instead of reconstructing
// "http://"+host.
func TestCDKCommandEndpointURLHTTPSNoDockerRequired(t *testing.T) {
	t.Parallel()
	requireLinuxForSSLCertFileTrust(t)

	srv := awsHealthTLSServer(t)
	defer srv.Close()

	fakeDir := writeFakeCDK(t, "2.177.0")
	e := env.With(env.DisableEvents, "1").With("PATH", fakeDir).With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost, httpsCertFileEnv(t, srv))

	stdout, stderr, err := runLstk(t, testContext(t), t.TempDir(), e, "--endpoint-url", srv.URL, "cdk", "deploy", "MyStack")
	require.NoError(t, err, "stderr: %s", stderr)
	assert.Contains(t, stdout, "ENV_AWS_ENDPOINT_URL="+srv.URL)
}

// TestSnapshotSaveEndpointURLHTTPSNoDockerRequired proves the emulator-client
// layer (internal/emulator/aws/client.go's ExportState) reaches an
// externally-managed https endpoint directly, without downgrading to http
// (task 7.3).
func TestSnapshotSaveEndpointURLHTTPSNoDockerRequired(t *testing.T) {
	t.Parallel()
	requireLinuxForSSLCertFileTrust(t)

	var gotPath string
	srv := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/_localstack/health":
			w.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(w).Encode(map[string]any{
				"version":  "3.0.2",
				"services": map[string]string{"s3": "available"},
			})
		case "/_localstack/pods/state":
			gotPath = r.URL.Path
			w.Header().Set("Content-Type", "application/zip")
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write(minimalStateZip(t))
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	defer srv.Close()

	dir := t.TempDir()
	e := env.With(env.DisableEvents, "1").With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost, httpsCertFileEnv(t, srv))

	stdout, stderr, err := runLstk(t, testContext(t), dir, e, "--endpoint-url", srv.URL, "snapshot", "save")
	require.NoError(t, err, "stderr: %s", stderr)
	assert.Contains(t, stdout, "Snapshot saved")
	assert.Equal(t, "/_localstack/pods/state", gotPath)
}

// TestStatusEndpointURLHTTPSchemeMismatchSuggestsHTTPS covers the confusing
// error a user hit when pointing at a cloud-hosted instance with `http://`:
// nothing listens on port 80 there, so the only thing reported was a raw "no
// route to host" that never hinted the scheme was the problem. lstk now
// re-probes the other scheme on the failure path and names the URL to retry.
func TestStatusEndpointURLHTTPSchemeMismatchSuggestsHTTPS(t *testing.T) {
	t.Parallel()
	requireLinuxForSSLCertFileTrust(t)

	srv := awsHealthTLSServer(t)
	defer srv.Close()

	httpURL := strings.Replace(srv.URL, "https://", "http://", 1)
	e := env.With(env.DisableEvents, "1").With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost, httpsCertFileEnv(t, srv))

	_, stderr, err := runLstk(t, testContext(t), t.TempDir(), e, "--endpoint-url", httpURL, "status")
	require.Error(t, err)
	assert.Contains(t, stderr, "could not reach LocalStack emulator at "+httpURL)
	assert.Contains(t, stderr, srv.URL+" responded")
	assert.Contains(t, stderr, "retry with that URL")
	assert.NotContains(t, stderr, "lstk start")
}

// TestStatusEndpointURLHTTPSSchemeMismatchSuggestsHTTP is the mirror image:
// an `https://` URL aimed at a plain-HTTP emulator fails on the TLS handshake,
// which is just as opaque, and gets the same treatment. No certificate trust
// is involved, so unlike its siblings here this one runs on every platform.
func TestStatusEndpointURLHTTPSSchemeMismatchSuggestsHTTP(t *testing.T) {
	t.Parallel()

	srv := awsHealthServer(t)
	defer srv.Close()

	httpsURL := strings.Replace(srv.URL, "http://", "https://", 1)
	e := env.With(env.DisableEvents, "1").With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost)

	_, stderr, err := runLstk(t, testContext(t), t.TempDir(), e, "--endpoint-url", httpsURL, "status")
	require.Error(t, err)
	assert.Contains(t, stderr, "could not reach LocalStack emulator at "+httpsURL)
	assert.Contains(t, stderr, srv.URL+" responded")
	assert.Contains(t, stderr, "retry with that URL")
}

// TestStatusEndpointURLHTTPSRendersReducedOutput is the https counterpart of
// TestStatusEndpointURLRendersReducedOutput (task 7.5).
func TestStatusEndpointURLHTTPSRendersReducedOutput(t *testing.T) {
	t.Parallel()
	requireLinuxForSSLCertFileTrust(t)

	srv := awsHealthTLSServer(t)
	defer srv.Close()

	e := env.With(env.DisableEvents, "1").With(env.Home, t.TempDir())
	e = append(e, unreachableDockerHost, httpsCertFileEnv(t, srv))

	stdout, stderr, err := runLstk(t, testContext(t), t.TempDir(), e, "--endpoint-url", srv.URL, "status")
	require.NoError(t, err, "stderr: %s", stderr)
	assert.Contains(t, stdout, "running")
	assert.Contains(t, stdout, srv.URL)
	assert.Contains(t, stdout, "3.0.2")
	assert.NotContains(t, stdout, "Container:")
	assert.NotContains(t, stdout, "Uptime:")
}

File: test/e2e/tests/endpoint-url-https.test.ts
import { execa } from "execa";
import { describe, expect, test } from "vitest";
import { lstk, requirement, tempHome, unreachableDockerHost } from "../support/index.ts";
import { emulatorApi, sslCertFileTrusted, stateExportZip } from "../support/emulator-api.ts";
import { fakeBinary } from "../support/fake-binary.ts";

// Ported from test/integration/endpoint_url_https_test.go.
//
// A LocalStack cloud-hosted ephemeral instance is reached over https, so the
// resolved scheme has to survive all the way to whatever finally makes the
// request — the wrapped aws/cdk tools and the emulator API calls behind
// snapshot/status — instead of being reconstructed as "http://"+host.
//
// These need lstk to trust a throwaway certificate, and lstk has no --insecure
// flag by design. The only handle is the OS trust mechanism its TLS stack reads,
// and Go's crypto/x509 honours SSL_CERT_FILE on Linux only: darwin goes through
// Security.framework and Windows through its own store. So all but the last test
// here are Linux-only, exactly as they were in Go.

const noOpenssl = requirement(
  "the openssl CLI",
  await execa("openssl", ["version"], { reject: false }).then((r) => r.exitCode === 0),
  "Install openssl, used to mint a throwaway certificate for the https mock.",
);
const noCertTrust = requirement(
  "SSL_CERT_FILE certificate trust",
  sslCertFileTrusted,
  "Run on Linux: Go's x509 verifier only reads SSL_CERT_FILE there, so an exec'd lstk cannot be told to trust the test certificate on macOS or Windows.",
);

/** Env that both breaks Docker and makes lstk trust the mock's certificate. */
function tlsEnv(certFile: string): Record<string, string> {
  return { DOCKER_HOST: unreachableDockerHost, SSL_CERT_FILE: certFile };
}

describe.skipIf(noOpenssl || noCertTrust)("an https endpoint", () => {
  test("reaches the wrapped aws CLI with the https scheme intact", async () => {
    const api = await emulatorApi({ version: "3.0.2", tls: true });
    const aws = await fakeBinary({ name: "aws" });
    const home = await tempHome();

    const run = await lstk(["--endpoint-url", api.url, "aws", "s3", "ls"], {
      home,
      env: { ...tlsEnv(api.certFile!), PATH: aws.path },
    });

    expect(run).toSucceed();
    expect((await aws.lastCall())?.args).toEqual(["--endpoint-url", api.url, "s3", "ls"]);
  });

  test("reaches the wrapped cdk with the https scheme intact", async () => {
    const api = await emulatorApi({ version: "3.0.2", tls: true });
    const cdk = await fakeBinary({
      name: "cdk",
      responses: [{ when: ["--version"], stdout: "2.177.0\n" }],
    });
    const home = await tempHome();

    const run = await lstk(["--endpoint-url", api.url, "cdk", "deploy", "MyStack"], {
      home,
      env: { ...tlsEnv(api.certFile!), PATH: cdk.path },
    });

    expect(run).toSucceed();
    expect((await cdk.lastCall())?.env.AWS_ENDPOINT_URL).toBe(api.url);
  });

  // The emulator-client layer, rather than a wrapped tool: `snapshot save` calls
  // /_localstack/pods/state itself.
  test("is where snapshot save exports state from", async () => {
    const api = await emulatorApi({ version: "3.0.2", tls: true, stateExport: stateExportZip() });
    const home = await tempHome();

    const run = await lstk(["--endpoint-url", api.url, "snapshot", "save"], {
      home,
      env: tlsEnv(api.certFile!),
      cwd: home.path,
    });

    expect(run).toSucceed();
    expect(run).toPrint("Snapshot saved");
    expect(api.requestedPaths()).toContain("/_localstack/pods/state");
  });

  test("renders reachability, type and version, with no container fields", async () => {
    const api = await emulatorApi({ version: "3.0.2", services: { s3: "available" }, tls: true });
    const home = await tempHome();

    const run = await lstk(["--endpoint-url", api.url, "status"], {
      home,
      env: tlsEnv(api.certFile!),
    });

    expect(run).toSucceed();
    expect(run.stdout).toPrintExactly(`
      Fetching LocalStack status...
      ✔︎ LocalStack AWS Emulator is running
      • Endpoint: ${api.url}
      • Version: 3.0.2
      > Note: No resources deployed
    `);
  });

  // The confusing case a user actually hit: nothing listens on port 80 of a
  // cloud-hosted instance, so an `http://` URL produced a bare "no route to
  // host" that never hinted the scheme was wrong.
  test("is suggested by name when the user reached for http", async () => {
    const api = await emulatorApi({ version: "3.0.2", tls: true });
    const httpUrl = api.url.replace("https://", "http://");
    const home = await tempHome();

    const run = await lstk(["--endpoint-url", httpUrl, "status"], {
      home,
      env: tlsEnv(api.certFile!),
    });

    expect(run).toExitWith(1);
    expect(run).toPrint(`could not reach LocalStack emulator at ${httpUrl}`);
    expect(run).toPrint(`${api.url} responded`);
    expect(run).toPrint("retry with that URL");
    expect(run, "the user manages this endpoint themselves").not.toPrint("lstk start");
  });
});
